File: 150_Evaluate_Reverse_Polish_Notation/xsy/solution.py
class Solution:
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []
        operations = {
            "+": "ADD",
            "-": "SUB",
            "*": "MUL",
            "/": "DIV",
        }
        for token in tokens:
            if token not in operations:
                stack.append(int(token))
            else:
                operation = operations[token]
                number2, number1 = stack.pop(), stack.pop()
                if operation == "ADD":
                    number = number1 + number2
                elif operation == "SUB":
                    number = number1 - number2
                elif operation == "MUL":
                    number = number1 * number2
                else:
                    number = int(number1 / number2)
                stack.append(number)
        return stack.pop()


s = Solution()
print(s.evalRPN(["10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"]))

File: 150_Evaluate_Reverse_Polish_Notation/xsy/solution.go
func evalRPN(tokens []string) int {
	operations := map[string]func(int, int) int{
		"+": func(number1, number2 int) int {return number1+number2},
		"-": func(number1, number2 int) int {return number1-number2},
		"*": func(number1, number2 int) int {return number1*number2},
		"/": func(number1, number2 int) int {return int(number1/number2)},
	}
	stack := []int{}
	for _, token := range tokens {
		if operationfunc, had := operations[token]; had {
			length := len(stack)
			number1, number2 := stack[length-2], stack[length-1]
			stack = stack[:length-2]
			number := operationfunc(number1, number2)
			stack = append(stack, number)
		} else {
			number, _ := strconv.Atoi(token)
			stack = append(stack, number)
		}
	}
	return stack[0]
}
